File: Migraine/src/migraine.cpp
#include "migraine.h"

#include <iostream>

namespace br
{
	void migraine::gotoIndex(const int& intRegister)
	{
		setFunctionIndex(intRegister);
	}

	int migraine::number(const int& intRegister)
	{
		return (intRegister * 10) + (getRealFunctionIndex() % 10);
	}

	int migraine::add(const int& intRegister)
	{
		return intRegister + integerRegisters[!currentRegister];
	}

	int migraine::subtract(const int& intRegister)
	{
		return intRegister - integerRegisters[!currentRegister];
	}

	int migraine::multiply(const int& intRegister)
	{
		return intRegister * integerRegisters[!currentRegister];
	}

	int migraine::divide(const int& intRegister)
	{
		int value = integerRegisters[!currentRegister];
		return value != 0 ? intRegister / value : 0;
	}

	int migraine::input(const int& intRegister)
	{
		int readValue;
		std::cin >> readValue;
		return readValue;
	}

	int migraine::outputInt(const int& intRegister)
	{
		std::cout << intRegister;
		return intRegister;
	}

	int migraine::outputChar(const int& intRegister)
	{
		std::cout << (char)intRegister;
		return intRegister;
	}

	int migraine::compareEquals(const int& intRegister)
	{
		return intRegister == integerRegisters[!currentRegister];
	}

	int migraine::compareGreater(const int& intRegister)
	{
		return intRegister > integerRegisters[!currentRegister];
	}

	int migraine::branch(const int& intRegister)
	{
		int isBranching = (unsigned)intRegister > 0;
		if (isBranching)
		{
			int newProgramCounter = integerRegisters[!currentRegister];
			programCounter = newProgramCounter;
		}
		return intRegister;
	}

	int migraine::stackPush(const int& intRegister)
	{
		integerStack.push_back(intRegister);
		return intRegister;
	}

	int migraine::stackPop(const int& intRegister)
	{
		int value = 0;
		if (integerStack.size() > 0)
		{
			value = integerStack.back();
			integerStack.pop_back();
		}
		return value;
	}

	int migraine::stackRead(const int& intRegister)
	{
		int value = 0;
		if (intRegister < integerStack.size())
		{
			value = integerStack[intRegister];
		}
		return value;
	}

	int migraine::stackWrite(const int& intRegister)
	{
		if (intRegister < integerStack.size())
		{
			integerStack[intRegister] = integerRegisters[!currentRegister];
		}
		return intRegister;
	}

	void migraine::buildFunctionList()
	{
		for (int i = 0; i < 10; ++i)
		{
			functionList.push_back(&migraine::number);
		}

		functionList.push_back(&migraine::add);				// 10
		functionList.push_back(&migraine::subtract);		
		functionList.push_back(&migraine::multiply);		
		functionList.push_back(&migraine::divide);			

		functionList.push_back(&migraine::input);			// 14
		functionList.push_back(&migraine::outputInt);
		functionList.push_back(&migraine::outputChar);

		functionList.push_back(&migraine::compareEquals);	// 17
		functionList.push_back(&migraine::compareGreater);
		functionList.push_back(&migraine::branch);

		functionList.push_back(&migraine::stackPush);		// 20
		functionList.push_back(&migraine::stackPop);
		functionList.push_back(&migraine::stackRead);
		functionList.push_back(&migraine::stackWrite);

		functionList.shrink_to_fit();
	}

	void migraine::setFunctionIndex(const unsigned int& newIndex)
	{
		functionIndex = newIndex;
		if (getRealFunctionIndex() >= functionList.size())
		{
			functionIndex = 0;
		}
	}

	void migraine::runFunction(const int& intRegister)
	{
		bool isGotoFunction = ((functionIndex + 1) % 2) == 0;

		if (isGotoFunction)
		{
			gotoIndex(intRegister);
		}
		else
		{
			integerRegisters[currentRegister] = (this->*functionList[getRealFunctionIndex()])(intRegister);
		}
	}

	void migraine::run(std::string program)
	{
		size_t programLength = program.length();
		for (programCounter = 0; programCounter < programLength; programCounter++)
		{
			char currentChar = program.at(programCounter);
			
			if (currentChar == IndexIncrement)
			{
				setFunctionIndex(functionIndex + 1);
				continue;
			}

			if (currentChar == SwapRegister)
			{
				currentRegister = !currentRegister;
				continue;
			}

			if (currentChar == PassZeroFunction)
			{
				runFunction(0);
				continue;
			}

			if (currentChar == PipeFunction)
			{
				runFunction(integerRegisters[currentRegister]);
				continue;
			}
		}
	}

	migraine::migraine() : integerRegisters{ 0, 0 }, functionIndex{ 0 }, currentRegister{ 0 }
	{
		buildFunctionList();
		integerStack.reserve(65535);
	}
}

File: MigraineJS/src/migraine.js
export var migraineState = {
    version : "1.1",

    indexIncrement : 'e',
    passZeroFunction : '!',
    pipeFunction : '.',
    swapRegister : 'E',

    programCounter : 0,
    functionIndex : 0,
    integerStack : [],
    callStack : [],
    registers : [0, 0],
    currentRegister : 0,

    standardInput : "",
    standardOutput : "",
    callbackOutput : (output) => { return output; },

    functionList : []
}

function getRealFunctionIndex(migraineState) {
    return parseInt(migraineState.functionIndex / 2);
}

function setFunctionIndex(migraineState, register) {
    migraineState.functionIndex = parseInt(register);
    if (getRealFunctionIndex(migraineState) >= migraineState.functionList.length) {
        migraineState.functionIndex = 0;
    }
}

function getOppositeRegister(migraineState) {
    return migraineState.registers[migraineState.currentRegister === 0 ? 1 : 0];
}

function gotoIndex(migraineState, register) {
    setFunctionIndex(migraineState, register);
}
function number(migraineState, register) {
    return parseInt((register * 10) + getRealFunctionIndex(migraineState));
}
function add(migraineState, register) {
    return register + getOppositeRegister(migraineState);
}
function subtract(migraineState, register) {
    return register - getOppositeRegister(migraineState);
}
function multiply(migraineState, register) {
    return register * getOppositeRegister(migraineState);
}
function divide(migraineState, register) {
    return register / getOppositeRegister(migraineState);
}
function input(migraineState, register) {
    return migraineState.standardInput;
}
function outputInt(migraineState, register) {
    migraineState.standardOutput += parseInt(register).toString();
    migraineState.callbackOutput(migraineState.standardOutput);
    return register;
}
function outputChar(migraineState, register) {
    migraineState.standardOutput += String.fromCharCode(register);
    migraineState.callbackOutput(migraineState.standardOutput);
    return register;
}
function compareEquals(migraineState, register) {
    return register === getOppositeRegister(migraineState);
}
function compareGreater(migraineState, register) {
    return register > getOppositeRegister(migraineState);
}
function branch(migraineState, register) {
    let isBranching = register > 0;
    if (isBranching) {
        migraineState.callStack.push(migraineState.programCounter);
        migraineState.programCounter = getOppositeRegister(migraineState);
        migraineState.currentRegister = 0;
    }
    return register;
}
function branchReturn(migraineState, register) {
    let value = 0;
    if (migraineState.callStack.length > 0) {
        value = migraineState.callStack.pop();
    }
    migraineState.programCounter = register + value;
    return register;
}
function callStackPop(migraineState, register) {
    if (migraineState.callStack.length > 0) {
        migraineState.callStack.pop();
    }
    return register;
}
function stackPush(migraineState, register) {
    migraineState.integerStack.push(register);
    return register;
}
function stackPop(migraineState, register) {
    let value = 0;
    if (migraineState.integerStack.length > 0) {
        value = migraineState.integerStack.pop();
    }
    return value;
}
function stackRead(migraineState, register) {
    let value = 0;
    if (register < migraineState.integerStack.length) {
        value = migraineState.integerStack[register];
    }
    return value;
}
function stackWrite(migraineState, register) {
    if (register < migraineState.integerStack.length) {
        migraineState.integerStack[register] = getOppositeRegister(migraineState);
    }
    return register;
}

function initializeFunctionList(migraineState) {
    migraineState.functionList = [
        number, number, number, number, number, number, number, number, number, number,
        add,
        subtract,
        multiply,
        divide,
        input,
        outputInt,
        outputChar,
        compareEquals,
        compareGreater,
        branch,
        branchReturn,
        callStackPop,
        stackPush,
        stackPop,
        stackRead,
        stackWrite
    ];
}

function runFunction(migraineState, register) {
    let isGotoFunction = parseInt((migraineState.functionIndex + 1) % 2) === 0;
    if (isGotoFunction) {
        gotoIndex(migraineState, register);
    } else {
        migraineState.registers[migraineState.currentRegister] = migraineState.functionList.at(getRealFunctionIndex(migraineState))(migraineState, register);
    }
}

export function step(migraineState, program) {
    if (migraineState.programCounter >= program.length)
        return;

    let currentStep = program.charAt(migraineState.programCounter);

    if (currentStep === migraineState.indexIncrement) {
        setFunctionIndex(migraineState, migraineState.functionIndex + 1);
    }
    if (currentStep === migraineState.swapRegister) {
        migraineState.currentRegister = migraineState.currentRegister === 0 ? 1 : 0;
    }
    if (currentStep === migraineState.passZeroFunction) {
        runFunction(migraineState, 0);
    }
    if (currentStep === migraineState.pipeFunction) {
        runFunction(migraineState, migraineState.registers[migraineState.currentRegister]);
    }

    migraineState.programCounter += 1;
}

export function convertProgram (program) {
    return program.match(/([e!\.E]*)/g).join("");
}

export function run(migraineState, program) {
    let programLength = program.length;
    for (migraineState.programCounter = 0; migraineState.programCounter < programLength; migraineState.programCounter++) {
        let currentChar = program.charAt(migraineState.programCounter);
        step(migraineState, currentChar);
    }
}

export function initialize(migraineState) {
    initializeFunctionList(migraineState);
}

export function resetState(migraineState) {
    migraineState.programCounter = 0;
    migraineState.functionIndex = 0;
    migraineState.integerStack = [];
    migraineState.callStack = [];
    migraineState.registers = [0, 0];
    migraineState.currentRegister = 0;
    migraineState.functionList = [];

    migraineState.standardInput = "";
    migraineState.standardOutput = "";
}